tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -53,75 +66,110 @@ parseByteString language bytestring = withParser language $ \ parser -> withPars
   else
     withRootNode treePtr $ \ rootPtr ->
       withCursor (castPtr rootPtr) $ \ cursor ->
-        runM (runFail (runReader cursor (runReader bytestring (peekNode >>= unmarshalNode))))
+        (Right <$> runReader (UnmarshalState bytestring cursor) (liftIO (peek rootPtr) >>= unmarshalNode))
+          `catch` (pure . Left . getUnmarshalError)
+
+newtype UnmarshalError = UnmarshalError { getUnmarshalError :: String }
+  deriving (Show)
```

Comment:
Throwing exceptions in `IO` instead of threading them through via `FailC` was a minor optimization.